### DISCUSSION TITLE
Best way to debug unit tests in Java and JS?

### DISCUSSION THREAD

**Discussion by DragonAxe** [Vertispan/j2clmavenplugin#273] [Q&A] [answered]
What is the best way to debug my unit test on the JavaScript side (`mvn test`)?
How can I get `DomGlobal.console.log` output from a JavaScript test?
Is there a way I can run a test from an external browser by hosting the test code from the jetty server?

I can debug the Java side in Intellij by adding a `junit:junit:4.13.2` dependency to my project.
```
    <dependency>
      <groupId>com.vertispan.j2cl</groupId>
      <artifactId>junit-emul</artifactId>
      <version>${j2cl.version}</version>
      <classifier>sources</classifier>
      <scope>test</scope>
    </dependency>

    <dependency>
      <groupId>junit</groupId>
      <artifactId>junit</artifactId>
      <version>4.13.2</version>
      <scope>test</scope>
    </dependency>

<!-- I noticed the junit dependency and this dependency conflict with the following error in Intellij:
Exception in thread "main" java.lang.NoSuchMethodError: 'org.junit.runner.Description org.junit.runner.Runner.getDescription()'
	at com.intellij.junit4.JUnit4IdeaTestRunner.getDescription(JUnit4IdeaTestRunner.java:79)
-->
<!--    <dependency>-->
<!--      <groupId>com.vertispan.j2cl</groupId>-->
<!--      <artifactId>junit-emul</artifactId>-->
<!--      <version>${j2cl.version}</version>-->
<!--      <scope>test</scope>-->
<!--    </dependency>-->
```
Thanks in advance

**Comment by DragonAxe** [comment, marked as answer]
Ah, I discovered I can update the jetty configuration to point to the test build folder:
 `<resourceBase>${project.build.directory}/${project.build.finalName}-test</resourceBase>`
I can then point my browser to: `localhost:8080/ProjectName/test-com.example.ExampleTest.html` and debug from there.

**Comment by niloc132** [comment]
Glad you figured out the test directory.

Take care when changing the junit version - there is no runner that will let you invoke these tests from your IDE at this time anyway, so swapping that dependency probably has no benefit. It also might break things, by giving your client classpath a jar that may not be compatible with the junit emulation that exists for J2CL.

**Reply by DragonAxe** [reply]
When I try to run a test from Intellij with the following dependencies:
```
com.vertispan.j2cl:junit-emul:v20230718-1 (classifier: sources)
com.vertispan.j2cl:junit-emul:v20230718-1
(no junit dependency)
```
I get the following error:
```
!!! JUnit version 3.8 or later expected:
java.lang.NoClassDefFoundError: org/junit/runner/notification/RunListener
...
```
When you say "there is no runner that will let you invoke these tests from your IDE at this time", is the above what you are referring to?

**Reply by niloc132** [reply]
You cannot run j2cl:test from intellij, no matter how you set up dependencies.

J2CL (especially in maven, with its anemic classpath management) is really meant to have small projects set up for a specific classpath. If you are saying that you are mixing JVM tests with J2CL tests, I would encourage you to avoid that to prevent future headaches - split the project into "shared" code (for "this runs in jvm and j2cl") vs "client" ("this code only makes sense in a browser"), and probably do the jvm tests in shared, j2cl tests in client.

If you change the dependencies in order to avoid that NoClassDefFoundError, can your j2cl tests run in IJ? I assume not...

**Reply by DragonAxe** [reply]
I see now. It makes sense that individual J2cl tests can't be run from Intellij. Good idea to keep j2cl and jvm tests in separate projects. Above, I was trying to write a test that could pass in jvm and j2cl to see if it was possible.

Is there a better way to prevent j2cl tests from running in the jvm?
```
import com.google.j2cl.junit.apt.J2clTestInput;
import org.junit.Test;

@J2clTestInput(ComponentTest.class)
public class ComponentTest {
  @Test
  public void testNewObject() {
    if (DomGlobal.document != null) {
       // j2cl code here
```

Also, is there a way to tell `mvn test` to run a specific test class? It looks like each test class ~~has to be compiled~~ (Edit: Maybe I'm mistaken, see output below for what it's actually doing) separately which takes a few moments and could easily become minutes to run a handful of test classes. It's a little faster if I use compilationLevel=BUNDLE_JAR and webdriver=chrome because it skips the optimization step.

```
[INFO] --- j2cl:0.22.0:test (test-js) @ ExampleProject ---
[INFO] Removing dependency com.google.jsinterop:base:jar:1.0.1:compile, replacing with com.vertispan.jsinterop:base:jar:1.0.1-1
[INFO] Removing dependency com.google.jsinterop:base:jar:1.0.1:compile, replacing with com.vertispan.jsinterop:base:jar:1.0.1-1
[INFO] Removing dependency com.google.jsinterop:base:jar:1.0.1:compile, replacing with com.vertispan.jsinterop:base:jar:1.0.1-1
[INFO] Removing dependency com.google.jsinterop:base:jar:1.0.3:compile, replacing with com.vertispan.jsinterop:base:jar:1.0.1-1
[INFO] Removing dependency org.jspecify:jspecify:jar:0.3.0:compile, no replacement
[INFO] Starting final task com.example:ExampleProject:1.0-SNAPSHOT-test/test_summary
[INFO] Finished final task com.example:ExampleProject:1.0-SNAPSHOT-test/test_summary in 8ms
[INFO] Starting final task com.example:ExampleProject:1.0-SNAPSHOT-test-com/example/ComponentTest.js/bundled_js_app
[INFO] Finished final task com.example:ExampleProject:1.0-SNAPSHOT-test-com/example/ComponentTest.js/bundled_js_app in 395ms
[INFO] Test started: com.example.ComponentTest
```
